### ISSUE TITLE
Document Backends

### ISSUE THREAD

**Issue by ecki** [asciidoctor/asciidoctorj#438]
When using AsciidoctorJ (via Maven Plugin) I tried some of the standard Asciidoctor backends (xhtml11, html4, ...) and some of them did not work. It causes a rather strange error message

    org.asciidoctor.internal.JRubyAsciidoctor renderFile
    FATAL: (NoMethodError) asciidoctor: FAILED: src/main/asciidoc/main.adoc:
    Failed to load AsciiDoc document - undefined method `convert' for nil:NilClass

So it is hard to know what backends are provided, and what backends require additional extensions.

   `xhtml5`, `html5`, `html` - seems to work out of the box (but all is same html?)
   `docbook`, `docbook5` - seems to work out of the box
   `docbook45` - seems to work out of the box
   `pdf` - requires asciidocj-pdf
   `epub3` - requires asciidocj-epub3

Anyting more? (can `manpage`, `deckjs` and `latex` be used?)

**Comment by mojavelinux** [comment]
Please see http://asciidoctor.org/docs/user-manual/#selecting-an-output-format for a fairly complete overview of the available options.

Asciidoctor does not recognize xhtml11 or html4 as these are old standards that are no longer relevant in modern times (though, in theory, someone could create a converter to support them).

One group of backend names that are intentionally left out of the section I referenced are the HTML5 presentation systems (e.g., deckjs, revealjs, etc). I don't think they're ready for general usage yet because they lack releases. However, I am planning to release an alpha of the Bespoke.js converter (mapped to the backend name `bespoke`) in the next few days. That's the only HTML5 presentation system I'll be focusing on, and therefore will be the one most likely to be covered in the user manual.

**Comment by ecki** [comment]
Thanks for the pointer. At least `manpage` does not work for AsciidoctorJ Maven Plugin (at least I dont know how) and for latex and mallard I am not sure which maven dependency to use in the plugin.

What about the error message for unknown backends, should I open a seperate bug, if yes is this a problem of the AsciiDoctorJ library or of the maven plugin?

**Comment by mojavelinux** [comment]
>  for latex and mallard I am not sure which maven dependency to use in the plugin.

There is currently no release for these (in fact, mallard isn't released at all, as I mention). We still have to sort out a simpler way to use gems from the Java side. The correct fix is to get JRuby to be able to consume a gem natively. In the meantime, we've been redistributing gems as jar files to make it simpler (hence asciidoctorj-pdf, asciidoctorj-epub3, etc), but I consider this to be a hack.

**Comment by mojavelinux** [comment]
> What about the error message for unknown backends, should I open a seperate bug, if yes is this a problem of the AsciiDoctorJ library or of the maven plugin?

That seems to be an issue in AsciidoctorJ. In Asciidoctor Ruby, you'd get the message:

```
asciidoctor: FAILED: missing converter for backend 'bogus'. Processing aborted. (RuntimeError)
```

I'm not sure why AsciidoctorJ isn't throwing the same exception.

**Comment by mojavelinux** [comment]
> At least manpage does not work for AsciidoctorJ Maven Plugin

The manpage converter should work (if you are using AsciidoctorJ 1.5.4). Keep in mind, however, that the manpage converter will get substantially better in the upcoming 1.5.5, so it's definitely a work in progress / new feature.

**Comment by mojavelinux** [comment]
Assuming the gem is released, you can use the gem Maven plugin to install it as part of the build. Once that is done, you just need to add a requires for it. There is an example of this somewhere in the maven examples. See https://github.com/asciidoctor/asciidoctor-maven-examples/issues/41.

**Comment by ecki** [comment]
Yes, actually I plan to contribute a manpage example, it might take some time as I am not familiar with Ruby Gems.

**Comment by abelsromero** [comment]
`manpage` works in both maven plugin an asciidoctorj, but seems it expects some specific format. Not any AsciiDoc document will work.
I created a couple of examples:
* AsciidoctorJ: https://github.com/abelsromero/asciidoctorj-gradle/blob/master/src/main/java/org/asciidoctor/demos/ManpageRenderer.java. Note it only works with this document: https://github.com/abelsromero/asciidoctorj-gradle/blob/master/src/asciidoc/manpage.adoc
* Asciidoctor-maven-plugin: https://github.com/abelsromero/asciidoctor-maven-examples/blob/manpage-example/asciidoc-multiple-inputs-example/pom.xml#L98-L109. Same as before, it parses the asciidoctor manpage, other documents may fail with some nasty errors.

Here i found some more info: http://asciidoctor.org/docs/user-manual/#man-pages

**Comment by ecki** [comment]
Yes you are right, `manpage` works I just tried it again. I had by accident defined a syntax highliter and when you do, you get a similiar cryptic error like when the backend is missing:

When this is added:

                               <sourceHighlighter>coderay</sourceHighlighter>

I see

    [ERROR] Failed to execute goal org.asciidoctor:asciidoctor-maven-plugin:1.5.2:process-asciidoc (output-man)
        on project asciidoc-demo: Execution output-man of goal org.asciidoctor:asciidoctor-maven-plugin:1.5.2:process-asciidoc failed:
        org.jruby.exceptions.RaiseException: (NoMethodError) undefined method `empty?' for nil:NilClass -> [Help 1]

**Comment by mojavelinux** [comment]
> but seems it expects some specific format.

Correct. Due to the highly-specific format of a manpage, it requires that you adhere to a specific structure when making the AsciiDoc file.

You can use the manpage from Asciidoctor as an example. See https://raw.githubusercontent.com/asciidoctor/asciidoctor/master/man/asciidoctor.adoc

**Comment by ecki** [comment]
The format is not a problem as the error messages cleanly describe the elements missing. My problem was the Ruby exception when it is configured incorrectly. I will close this ticket since all points are addressed. Let me know where I can open the more specific bug reports on the two exceptions (instead of messages "backend unknoen" and "manpage backend does not support syntax highliter". I guess both is upstream?

**Comment by mojavelinux** [comment]
> backend unknown

https://github.com/asciidoctor/asciidoctorj/issues

> manpage backend does not support syntax highlighter

could you put the sample AsciiDoc file in a gist.github.com because I can't figure out how to reproduce your scenario. Once I can verify, that would likely go in the main Asciidoctor repository (https://github.com/asciidoctor/asciidoctor/issues).

**Comment by mojavelinux** [comment]
I believe you, I just need to see where it is going wrong.

**Comment by ecki** [comment]
The exception happens before the asciidoc file is read, so any one would do. But I will create a reproducer to make sure I am right :)

**Comment by abelsromero** [comment]
I have a note about documentation. `manpage` appears as both doctype and backend, and I've tested that both options work.
Is there a recomendation?

**Comment by robertpanzer** [comment]
I just made a test with current master trying to use the backend `Whatever`.

This message gets logged over JUL:
```
Mär 06, 2016 3:35:37 PM org.asciidoctor.internal.JRubyAsciidoctor renderFile
SCHWERWIEGEND: (RuntimeError) asciidoctor: FAILED: missing converter for backend 'Whatever'. Processing aborted.
```

And the exception looks like this:
```
org.asciidoctor.internal.AsciidoctorCoreException: org.jruby.exceptions.RaiseException: (RuntimeError) asciidoctor: FAILED: missing converter for backend 'Whatever'. Processing aborted.
```

What it the expectation?

**Comment by mojavelinux** [comment]
Thanks for checking Robert!

That the result I expect. That raises the question, how is @ecki getting a different result when using the Maven plugin? An older version of AsciidoctorJ? Something else?

**Comment by mojavelinux** [comment]
Here's what I get when using the Maven plugin with a bogus backend name:

```
[ERROR] Failed to execute goal org.asciidoctor:asciidoctor-maven-plugin:1.5.3:process-asciidoc (asciidoc-to-html) on project asciidoc-to-html-example: Execution asciidoc-to-html of goal org.asciidoctor:asciidoctor-maven-plugin:1.5.3:process-asciidoc failed: org.jruby.exceptions.RaiseException: (RuntimeError) asciidoctor: FAILED: missing converter for backend 'foo'. Processing aborted. -> [Help 1]
```

That's what I expect.

**Comment by mojavelinux** [comment]
I'm not able to reproduce the problem of using the manpage backend with the sourceHighlighter set.

**Comment by beikov** [comment]
I'm using a deckjs backend and when updating to asciidoctorj 1.5.6 I suddenly get errors like `asciidoctor-diagram: ERROR: Failed to generate image: can't modify frozen`. I guess this is related?
Here is a GitHub repository with a project that shows the problem: https://github.com/Blazebit/blaze-persistence-presentation
When setting the `asciidoctor.version` in `/feature-overview/presentation/pom.xml` to `1.5.6`, the build via `mvn process-resources` produces the error.

**Comment by beikov** [comment]
Ah ok, with 1.5.8 it works now, but updating JRuby to `9.1.16.0` fails now with `Failed to load AsciiDoc document - undefined method `start_with?' for nil:NilClass`, but I guess that's unrelated?

**Comment by robertpanzer** [comment]
I just had a look at the project and the setup for the asciidoctor-maven-plugin looks a bit strange to me.
Additional dependencies for asciidoctor should go into the configuration of the asciidoctor-maven-plugin instead of being plain project dependencies.
asciidoctor-diagram is already available as a ready to use dependency for asciidoctorj as asciidoctorj-diagram.
The configuration of the asciidoctor-maven-plugin looked like this for me:
```xml
            <plugin>
                <groupId>org.asciidoctor</groupId>
                <artifactId>asciidoctor-maven-plugin</artifactId>
                <version>${asciidoctor.version}</version>
                <configuration>
                    <gemPath>${project.build.directory}/gems-provided</gemPath>
                    <requires>
                        <require>asciidoctor-diagram</require>
                    </requires>
                </configuration>
                <executions>
                    <execution>
                        <id>generate-slides</id>
                        <phase>generate-resources</phase>
                        <goals>
                            <goal>process-asciidoc</goal>
                        </goals>
                        <configuration>
                            <sourceDirectory>src/main/slides</sourceDirectory>
                            <templateDir>${templateDir}</templateDir>
                            <outputDirectory>${project.slides.directory}</outputDirectory>
                            <backend>${backend}</backend>
                            <sourceHighlighter>prettify</sourceHighlighter>
                        </configuration>
                    </execution>
                </executions>
                <dependencies>
                    <dependency>
                        <groupId>org.asciidoctor</groupId>
                        <artifactId>asciidoctorj</artifactId>
                        <version>1.5.6</version>
                        <exclusions>
                          <exclusion>
                            <groupId>org.jruby</groupId>
                            <artifactId>jruby-complete</artifactId>
                          </exclusion>
                        </exclusions>
                    </dependency>
                    <dependency>
                        <groupId>org.jruby</groupId>
                        <artifactId>jruby-complete</artifactId>
                        <version>9.1.15.0</version>
                    </dependency>
                    <dependency>
                        <groupId>org.asciidoctor</groupId>
                        <artifactId>asciidoctorj-diagram</artifactId>
                        <version>1.5.8</version>
                    </dependency>
                </dependencies>
            </plugin>
        </plugins>
```
There should be no need to invoke the gem-maven-plugin yourself for asciidoctor-diagram.

With this the slides where created for me with Java 8.

(On Java 9 I had a deadlock though, which is a bit strange.)

**Comment by mojavelinux** [comment]
> There should be no need to invoke the gem-maven-plugin yourself for asciidoctor-diagram.

I think that's a carry-over from before asciidoctorj-diagram was available.
